File: leetcode/543. Diameter of Binary Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int diameterOfBinaryTree(TreeNode* root) {
        auto tp = helper(root);
        return get<0>(tp) - 1;
    }
    
private:
    std::tuple<size_t, size_t> helper(TreeNode* node) {
        if (node == nullptr) {
            return make_tuple(0, 0);
        }
        
        auto ltp = helper(node->left);
        auto rtp = helper(node->right);
        
        size_t max_len = max({get<0>(ltp), get<0>(rtp), get<1>(ltp) + get<1>(rtp) + 1});
        size_t max_branch_len = 1 + max(get<1>(ltp), get<1>(rtp));
        
        return make_tuple(max_len, max_branch_len);
    } 
};

File: leetcode/tree/543. Diameter of Binary Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        res = 0
        def max_depth(node):
            if node is None:
                return 0
            left = max_depth(node.left)
            right = max_depth(node.right)
            nonlocal res
            res = max(res, left + right)
            return max(left, right) + 1

        max_depth(root)
        return res

